File: android/src/main/java/com/emallstudio/appmetrica_sdk/AppmetricaSdkPlugin.java
// Copyright 2019 EM ALL iT Studio. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.emallstudio.appmetrica_sdk;

import androidx.annotation.NonNull;

import android.util.Log;
import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;

import android.util.SparseArray;

import java.util.Arrays;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.view.FlutterView;

import com.yandex.metrica.YandexMetrica;
import com.yandex.metrica.YandexMetricaConfig;
import com.yandex.metrica.profile.Attribute;
import com.yandex.metrica.profile.StringAttribute;
import com.yandex.metrica.profile.UserProfile;
import com.yandex.metrica.profile.UserProfileUpdate;
import com.yandex.metrica.ecommerce.ECommerceScreen;
import com.yandex.metrica.ecommerce.ECommerceEvent;
import com.yandex.metrica.ecommerce.ECommerceAmount;
import com.yandex.metrica.ecommerce.ECommercePrice;
import com.yandex.metrica.ecommerce.ECommerceProduct;
import com.yandex.metrica.ecommerce.ECommerceCartItem;
import com.yandex.metrica.ecommerce.ECommerceReferrer;


/** AppmetricaSdkPlugin */
public class AppmetricaSdkPlugin implements MethodCallHandler, FlutterPlugin {
    private static final String TAG = "AppmetricaSdkPlugin";
    private MethodChannel methodChannel;
    private Context context;
    private Application application;

    /** Plugin registration for v1 embedder. */
    public static void registerWith(Registrar registrar) {
        final AppmetricaSdkPlugin instance = new AppmetricaSdkPlugin();
        instance.onAttachedToEngine(registrar.context(), registrar.messenger());
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
      onAttachedToEngine(flutterPluginBinding.getApplicationContext(), flutterPluginBinding.getBinaryMessenger());
    }

    private void onAttachedToEngine(Context applicationContext, BinaryMessenger binaryMessenger) {
      application = (Application) applicationContext;
      context = applicationContext;
      methodChannel = new MethodChannel(binaryMessenger, "emallstudio.com/appmetrica_sdk");
      methodChannel.setMethodCallHandler(this);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        methodChannel.setMethodCallHandler(null);
        methodChannel = null;
        context = null;
        application = null;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "activate":
                handleActivate(call, result);
                break;
            case "reportEvent":
                handleReportEvent(call, result);
                break;
            case "reportUserProfileCustomString":
                handleReportUserProfileCustomString(call, result);
                break;
            case "reportUserProfileCustomNumber":
                handleReportUserProfileCustomNumber(call, result);
                break;
            case "reportUserProfileCustomBoolean":
                handleReportUserProfileCustomBoolean(call, result);
                break;
            case "reportUserProfileCustomCounter":
                handleReportUserProfileCustomCounter(call, result);
                break;
            case "reportUserProfileUserName":
                handleReportUserProfileUserName(call, result);
                break;
            case "reportUserProfileNotificationsEnabled":
                handleReportUserProfileNotificationsEnabled(call, result);
                break;
            case "setStatisticsSending":
                handleSetStatisticsSending(call, result);
                break;
            case "getLibraryVersion":
                handleGetLibraryVersion(call, result);
                break;
            case "setUserProfileID":
                handleSetUserProfileID(call, result);
                break;
            case "sendEventsBuffer":
                handleSendEventsBuffer(call, result);
                break;
            case "reportReferralUrl":
                handleReportReferralUrl(call, result);
                break;
            case "reportEcommerceShowScreen":
                handleReportEcommerceShowScreen(call, result);
                break;
            case "reportEcommerceAddCart":
                handleReportEcommerceAddCart(call, result);
                break;
            default:
              result.notImplemented();
              break;
          }
    }

    private void handleActivate(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            // Get activation parameters.
            final String apiKey = (String) arguments.get("apiKey");
            final int sessionTimeout = (int) arguments.get("sessionTimeout");
            final boolean locationTracking = (boolean) arguments.get("locationTracking");
            final boolean statisticsSending = (boolean) arguments.get("statisticsSending");
            final boolean crashReporting = (boolean) arguments.get("crashReporting");
            final int maxReportsInDatabaseCount = (int) arguments.get("maxReportsInDatabaseCount");
            // Creating an extended library configuration.
            YandexMetricaConfig config = YandexMetricaConfig.newConfigBuilder(apiKey)
                    .withLogs()
                    .withSessionTimeout(sessionTimeout)
                    .withLocationTracking(locationTracking)
                    .withStatisticsSending(statisticsSending)
                    .withCrashReporting(crashReporting)
                    .withMaxReportsInDatabaseCount(maxReportsInDatabaseCount)
                    .build();
            // Initializing the AppMetrica SDK.
            YandexMetrica.activate(context, config);
            // Automatic tracking of user activity.
            YandexMetrica.enableActivityAutoTracking(application);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error performing activation", e.getMessage(), null);
        }
        result.success(null);
    }

    private void handleReportEvent(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String eventName = (String) arguments.get("name");
            @SuppressWarnings("unchecked")
            Map<String, Object> attributes = (Map<String, Object>) arguments.get("attributes");
            if (attributes == null) {
                YandexMetrica.reportEvent(eventName);
            } else {
                YandexMetrica.reportEvent(eventName, attributes);
            }

        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing event", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomString(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            final String value = (String) arguments.get("value");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (value != null) {
                profileBuilder.apply(Attribute.customString(key).withValue(value));
            } else {
                profileBuilder.apply(Attribute.customString(key).withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom string", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomNumber(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("value") != null) {
                final double value = (double) arguments.get("value");
                profileBuilder.apply(Attribute.customNumber(key).withValue(value));
            } else {
                profileBuilder.apply(Attribute.customNumber(key).withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom number", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomBoolean(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("value") != null) {
                final boolean value = (boolean) arguments.get("value");
                profileBuilder.apply(Attribute.customBoolean(key).withValue(value));
            } else {
                profileBuilder.apply(Attribute.customBoolean(key).withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom boolean", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileCustomCounter(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String key = (String) arguments.get("key");
            final double delta = (double) arguments.get("delta");
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            profileBuilder.apply(Attribute.customCounter(key).withDelta(delta));
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile custom counter", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileUserName(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("userName") != null) {
                final String userName = (String) arguments.get("userName");
                profileBuilder.apply(Attribute.name().withValue(userName));
            } else {
                profileBuilder.apply(Attribute.name().withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile user name", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportUserProfileNotificationsEnabled(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            UserProfile.Builder profileBuilder = UserProfile.newBuilder();
            if (arguments.get("notificationsEnabled") != null) {
                final boolean notificationsEnabled = (boolean) arguments.get("notificationsEnabled");
                profileBuilder.apply(Attribute.notificationsEnabled().withValue(notificationsEnabled));
            } else {
                profileBuilder.apply(Attribute.notificationsEnabled().withValueReset());
            }
            YandexMetrica.reportUserProfile(profileBuilder.build());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error reporing user profile user name", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleSetStatisticsSending(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final boolean statisticsSending = (boolean) arguments.get("statisticsSending");
            YandexMetrica.setStatisticsSending(context, statisticsSending);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error enable sending statistics", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleGetLibraryVersion(MethodCall call, Result result) {
        try {
            result.success(YandexMetrica.getLibraryVersion());
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error enable sending statistics", e.getMessage(), null);
        }
    }

    private void handleSetUserProfileID(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String userProfileID = (String) arguments.get("userProfileID");
            YandexMetrica.setUserProfileID(userProfileID);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error sets the ID of the user profile", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleSendEventsBuffer(MethodCall call, Result result) {
        try {
            YandexMetrica.sendEventsBuffer();
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error sending stored events from the buffer", e.getMessage(), null);

        }

        result.success(null);
    }

    private void handleReportReferralUrl(MethodCall call, Result result) {
        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String referral = (String) arguments.get("referral");
            YandexMetrica.reportReferralUrl(referral);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error sets the ID of the user profile", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportEcommerceShowScreen(MethodCall call, Result result) {
        try {
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String screenName = (String) arguments.get("screenName");
            // Creating a screen object. 
            ECommerceScreen screen = new ECommerceScreen()
                .setName(screenName);
            ECommerceEvent showScreenEvent = ECommerceEvent.showScreenEvent(screen);

            YandexMetrica.reportECommerce(showScreenEvent);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error report of e-commerce show screen", e.getMessage(), null);
        }

        result.success(null);
    }

    private void handleReportEcommerceAddCart(MethodCall call, Result result) {
        try {
            Map<String, Object> arguments = (Map<String, Object>) call.arguments;
            final String screenName = (String) arguments.get("screenName");
            // Creating a screen object. 
            ECommerceScreen screen = new ECommerceScreen()
                .setName(screenName);


            final String sku = (String) arguments.get("SKU");
            final String name = (String) arguments.get("name");
            final String category = (String) arguments.get("category");
            final double price = (double) arguments.get("price");
            final String reffer = (String) arguments.get("reffer");

            ECommercePrice originalPrice = new ECommercePrice(new ECommerceAmount(price, "RUB"));

            ECommerceProduct product = new ECommerceProduct(sku)
                .setOriginalPrice(originalPrice)
                .setName(name)
                .setCategoriesPath(Arrays.asList(category));

                ECommerceReferrer referrer = new ECommerceReferrer()
                    .setType(reffer) // Optional.
                    .setScreen(screen);

                    // Creating a cartItem object. 
             ECommerceCartItem addedItems1 = new ECommerceCartItem(product, originalPrice, 1.0)
                .setReferrer(referrer); // Optional.
            ECommerceEvent addCartItemEvent = ECommerceEvent.addCartItemEvent(addedItems1);

            YandexMetrica.reportECommerce(addCartItemEvent);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
            result.error("Error report of add cart", e.getMessage(), null);
        }

        result.success(null);
    }
}

File: ios/Classes/AppmetricaSdkPlugin.m
// Copyright 2019 EM ALL iT Studio. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "AppmetricaSdkPlugin.h"
#import <YandexMobileMetrica/YandexMobileMetrica.h>
#import <YandexMobileMetrica/YMMProfileAttribute.h>

@implementation AppmetricaSdkPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"emallstudio.com/appmetrica_sdk"
            binaryMessenger:[registrar messenger]];
  AppmetricaSdkPlugin* instance = [[AppmetricaSdkPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"activate" isEqualToString:call.method]) {
      [self handleActivate:call result:result];
  } else if ([@"reportEvent" isEqualToString:call.method]) {
      [self handleReportEvent:call result:result];
  } else if ([@"reportUserProfileCustomString" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomString:call result:result];
  } else if ([@"reportUserProfileCustomNumber" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomNumber:call result:result];
  } else if ([@"reportUserProfileCustomBoolean" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomBoolean:call result:result];
  } else if ([@"reportUserProfileCustomCounter" isEqualToString:call.method]) {
      [self handleReportUserProfileCustomCounter:call result:result];
  } else if ([@"reportUserProfileUserName" isEqualToString:call.method]) {
      [self handleReportUserProfileUserName:call result:result];
  } else if ([@"reportUserProfileNotificationsEnabled" isEqualToString:call.method]) {
      [self handleReportUserProfileNotificationsEnabled:call result:result];
  } else if ([@"setStatisticsSending" isEqualToString:call.method]) {
      [self handleSetStatisticsSending:call result:result];
  } else if ([@"getLibraryVersion" isEqualToString:call.method]) {
      [self handleGetLibraryVersion:call result:result];
  } else if ([@"setUserProfileID" isEqualToString:call.method]) {
      [self handleSetUserProfileID:call result:result];
  } else if ([@"sendEventsBuffer" isEqualToString:call.method]) {
      [self handleSendEventsBuffer:call result:result];
  } else if ([@"reportReferralUrl" isEqualToString:call.method]) {
      [self handleReportReferralUrl:call result:result];
  } else if ([@"reportEcommerceShowScreen" isEqualToString:call.method]) {
      [self handleReportEcommerceShowScreen:call result:result];
  } else if ([@"reportEcommerceAddCart" isEqualToString:call.method]) {
      [self handleReportEcommerceAddCart:call result:result];
  } else if ([@"showProductDetailsEventWithProduct" isEqualToString:call.method]) {
      [self handleShowProductDetailsEventWithProduct:call result:result];
  } else if ([@"reportEcommerceRemoveCart" isEqualToString:call.method]) {
      [self handleReportEcommerceRemoveCart:call result:result];
  } else if ([@"beginCheckoutEventWithOrder" isEqualToString:call.method]) {
      [self handleBeginCheckoutEventWithOrder:call result:result];
  }  else if ([@"purchaseEventWithOrder" isEqualToString:call.method]) {
      [self handlePurchaseEventWithOrder:call result:result];
  }  else {
      result(FlutterMethodNotImplemented);
  }
}

- (FlutterError* _Nullable)getFlutterError:(NSError*)error {
  if (error == nil) return nil;

  return [FlutterError errorWithCode:[NSString stringWithFormat:@"Error %ld", (long)error.code]
                             message:error.domain
                             details:error.localizedDescription];
}

- (void)handleActivate:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* apiKey = call.arguments[@"apiKey"];
    NSInteger sessionTimeout = [call.arguments[@"sessionTimeout"] integerValue];
    BOOL locationTracking = [call.arguments[@"locationTracking"] boolValue];
    BOOL statisticsSending = [call.arguments[@"statisticsSending"] boolValue];
    BOOL crashReporting = [call.arguments[@"crashReporting"] boolValue];
    NSInteger maxReportsInDatabaseCount = [call.arguments[@"maxReportsInDatabaseCount"] integerValue];
    // Creating an extended library configuration.
    YMMYandexMetricaConfiguration *configuration = [[YMMYandexMetricaConfiguration alloc] initWithApiKey:apiKey];
    // Setting up the configuration.
    configuration.logs = YES;
    configuration.sessionTimeout = sessionTimeout;
    configuration.locationTracking = locationTracking;
    configuration.statisticsSending = statisticsSending;
    configuration.crashReporting = crashReporting;
    // maxReportsInDatabaseCount is not supported yet. Android only feature.
    //configuration.maxReportsInDatabaseCount = maxReportsInDatabaseCount;
    // Initializing the AppMetrica SDK.
    [YMMYandexMetrica activateWithConfiguration:configuration];
    result(nil);
}

- (void)handleReportEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* name = call.arguments[@"name"];

    if (![call.arguments[@"attributes"] isEqual:[NSNull null]]) {
        NSDictionary* attributes = call.arguments[@"attributes"];
        [YMMYandexMetrica reportEvent:name
            parameters:attributes
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }
        ];
    } else {
        [YMMYandexMetrica reportEvent:name
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }
        ];
    }

    result(nil);
}


- (void)handleReportUserProfileCustomString:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"value"] isEqual:[NSNull null]]) {
        NSString* value = call.arguments[@"value"];
        [userProfile apply:[[YMMProfileAttribute customString:key] withValue:value]];
    } else {
        [userProfile apply:[[YMMProfileAttribute customString:key] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileCustomNumber:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"value"] isEqual:[NSNull null]]) {
        double value = [call.arguments[@"value"] doubleValue];
        [userProfile apply:[[YMMProfileAttribute customNumber:key] withValue:value]];
    } else {
        [userProfile apply:[[YMMProfileAttribute customNumber:key] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileCustomBoolean:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"value"] isEqual:[NSNull null]]) {
        BOOL value = [call.arguments[@"value"] boolValue];
        [userProfile apply:[[YMMProfileAttribute customBool:key] withValue:value]];
    } else {
        [userProfile apply:[[YMMProfileAttribute customBool:key] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileCustomCounter:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* key = call.arguments[@"key"];
    double value = [call.arguments[@"delta"] doubleValue];

    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    [userProfile apply:[[YMMProfileAttribute customCounter:key] withDelta:value]];
    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileUserName:(FlutterMethodCall*)call result:(FlutterResult)result {
    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"userName"] isEqual:[NSNull null]]) {
        NSString* userName = call.arguments[@"userName"];
        [userProfile apply:[[YMMProfileAttribute name] withValue:userName]];
    } else {
        [userProfile apply:[[YMMProfileAttribute name] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleReportUserProfileNotificationsEnabled:(FlutterMethodCall*)call result:(FlutterResult)result {
    YMMMutableUserProfile* userProfile =[ [YMMMutableUserProfile alloc] init];
    if (![call.arguments[@"notificationsEnabled"] isEqual:[NSNull null]]) {
        BOOL notificationsEnabled = [call.arguments[@"notificationsEnabled"] boolValue];
        [userProfile apply:[[YMMProfileAttribute notificationsEnabled] withValue:notificationsEnabled]];
    } else {
        [userProfile apply:[[YMMProfileAttribute notificationsEnabled] withValueReset]];
    }

    [YMMYandexMetrica reportUserProfile:userProfile
            onFailure:^(NSError *error) {
               result([self getFlutterError:error]);
            }];

    result(nil);
}

- (void)handleSetStatisticsSending:(FlutterMethodCall*)call result:(FlutterResult)result {
    BOOL statisticsSending = [call.arguments[@"statisticsSending"] boolValue];

    [YMMYandexMetrica setStatisticsSending:statisticsSending];

    result(nil);
}

- (void)handleGetLibraryVersion:(FlutterMethodCall*)call result:(FlutterResult)result {
    result([YMMYandexMetrica libraryVersion]);
}

- (void)handleSetUserProfileID:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* userProfileID = call.arguments[@"userProfileID"];

    [YMMYandexMetrica setUserProfileID:userProfileID];

    result(nil);
}

- (void)handleSendEventsBuffer:(FlutterMethodCall*)call result:(FlutterResult)result {
    [YMMYandexMetrica sendEventsBuffer];

    result(nil);
}

- (void)handleReportReferralUrl:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* referral = call.arguments[@"referral"];
    NSURL *url = [[NSURL alloc] initWithString:referral];
    
    [YMMYandexMetrica reportReferralUrl:url];

    result(nil);
}

/// show screen
-(void)handleReportEcommerceShowScreen:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* screenName = call.arguments[@"screenName"];
    YMMECommerceScreen *screen = [[YMMECommerceScreen alloc] initWithName:screenName];
    
    [YMMYandexMetrica reportECommerce:[YMMECommerce showScreenEventWithScreen:screen] onFailure:nil];
    
    result(nil);
}

-(void)handleReportEcommerceAddCart:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* screenName = call.arguments[@"screenName"];
    YMMECommerceScreen *screen = [[YMMECommerceScreen alloc] initWithName:screenName];
    
    NSString* SKU = call.arguments[@"SKU"];
    NSString* name = call.arguments[@"name"];
    NSString* category = call.arguments[@"category"];
    double price = [call.arguments[@"price"] doubleValue];
    NSString* reffer = call.arguments[@"reffer"];
    
    YMMECommerceAmount *actualFiat =
            [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:price]];
    
    YMMECommercePrice *originalPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
    
    YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:SKU
                                                                       name:name
                                                         categoryComponents:@[category]
                                                                    payload:@{}
                                                                actualPrice:originalPrice
                                                              originalPrice:originalPrice
                                                                 promoCodes:@[]];
    
    YMMECommerceReferrer *referrer = [[YMMECommerceReferrer alloc] initWithType:reffer
                                                                     identifier:@""
                                                                         screen:screen];
    
    NSDecimalNumber *quantity = [NSDecimalNumber decimalNumberWithString:@"1"];
    
    YMMECommerceCartItem *addedItems = [[YMMECommerceCartItem alloc] initWithProduct:product
                                                                        quantity:quantity
                                                                        revenue:originalPrice
                                                                        referrer:referrer];
    
    [YMMYandexMetrica reportECommerce:[YMMECommerce addCartItemEventWithItem:addedItems] onFailure:nil];
    
    result(nil);
}

-(void)handleReportEcommerceRemoveCart:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* screenName = call.arguments[@"screenName"];
    YMMECommerceScreen *screen = [[YMMECommerceScreen alloc] initWithName:screenName];
    
    NSString* SKU = call.arguments[@"SKU"];
    NSString* name = call.arguments[@"name"];
    NSString* category = call.arguments[@"category"];
    double price = [call.arguments[@"price"] doubleValue];
    NSString* reffer = call.arguments[@"reffer"];
    
    YMMECommerceAmount *actualFiat =
            [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:price]];
    
    YMMECommercePrice *originalPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
    
    YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:SKU
                                                                       name:name
                                                         categoryComponents:@[category]
                                                                    payload:@{}
                                                                actualPrice:originalPrice
                                                              originalPrice:originalPrice
                                                                 promoCodes:@[]];
    
    YMMECommerceReferrer *referrer = [[YMMECommerceReferrer alloc] initWithType:reffer
                                                                     identifier:@""
                                                                         screen:screen];
    
    NSDecimalNumber *quantity = [NSDecimalNumber decimalNumberWithString:@"1"];
    
    YMMECommerceCartItem *addedItems = [[YMMECommerceCartItem alloc] initWithProduct:product
                                                                        quantity:quantity
                                                                        revenue:originalPrice
                                                                        referrer:referrer];
    
    [YMMYandexMetrica reportECommerce:[YMMECommerce removeCartItemEventWithItem:addedItems] onFailure:nil];
    
    result(nil);
}

-(void)handleShowProductDetailsEventWithProduct:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    NSString* screenName = call.arguments[@"screenName"];
    YMMECommerceScreen *screen = [[YMMECommerceScreen alloc] initWithName:screenName];
    
    NSString* SKU = call.arguments[@"SKU"];
    NSString* name = call.arguments[@"name"];
    NSString* category = call.arguments[@"category"];
    double price = [call.arguments[@"price"] doubleValue];
    
    YMMECommerceAmount *actualFiat =
            [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:price]];
    
    YMMECommercePrice *originalPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
    
    YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:SKU
                                                                       name:name
                                                         categoryComponents:@[category]
                                                                    payload:@{}
                                                                actualPrice:originalPrice
                                                              originalPrice:originalPrice
                                                                 promoCodes:@[]];
    
    [YMMYandexMetrica reportECommerce:[YMMECommerce showProductCardEventWithProduct:product screen:screen] onFailure:nil];
    
    result(nil);
}

-(void)handleBeginCheckoutEventWithOrder:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    NSString* screenName = call.arguments[@"screenName"];
    YMMECommerceScreen *screen = [[YMMECommerceScreen alloc] initWithName:screenName];
    
    NSArray* productItems = call.arguments[@"products"];
    NSMutableArray<YMMECommerceCartItem *>* products = [[NSMutableArray alloc] init];
    
    NSDecimalNumber *quantity = [NSDecimalNumber decimalNumberWithString:@"1"];
    
    NSString* orderId = call.arguments[@"orderId"];
    
    for (NSDictionary* item in productItems) {
        NSString* SKU = item[@"SKU"];
        NSString* name = item[@"name"];
        NSString* category = item[@"category"];
        double price = [item[@"price"] doubleValue];
        NSString* reffer = item[@"reffer"];
        
        YMMECommerceAmount *actualFiat =
                [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:price]];
        
        YMMECommercePrice *originalPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
        
        YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:SKU
                                                                           name:name
                                                             categoryComponents:@[category]
                                                                        payload:@{}
                                                                    actualPrice:originalPrice
                                                                  originalPrice:originalPrice
                                                                     promoCodes:@[]];
        
        YMMECommerceReferrer *referrer = [[YMMECommerceReferrer alloc] initWithType:reffer
                                                                         identifier:@""
                                                                             screen:screen];
        
        YMMECommerceCartItem *addedItem = [[YMMECommerceCartItem alloc] initWithProduct:product
                                                                            quantity:quantity
                                                                            revenue:originalPrice
                                                                            referrer:referrer];
        
        [products addObject:addedItem];
    }
    
    
    YMMECommerceOrder *order = [[YMMECommerceOrder alloc] initWithIdentifier:orderId
                                                                   cartItems:products
                                                                     payload:@{}];
    
    [YMMYandexMetrica reportECommerce:[YMMECommerce beginCheckoutEventWithOrder:order] onFailure:nil];
    
    result(nil);
}

-(void)handlePurchaseEventWithOrder::(FlutterMethodCall*)call result:(FlutterResult)result {
    
    NSString* screenName = call.arguments[@"screenName"];
    YMMECommerceScreen *screen = [[YMMECommerceScreen alloc] initWithName:screenName];
    
    NSArray* productItems = call.arguments[@"products"];
    NSMutableArray<YMMECommerceCartItem *>* products = [[NSMutableArray alloc] init];
    
    NSDecimalNumber *quantity = [NSDecimalNumber decimalNumberWithString:@"1"];
    
    NSString* orderId = call.arguments[@"orderId"];
    
    for (NSDictionary* item in productItems) {
        NSString* SKU = item[@"SKU"];
        NSString* name = item[@"name"];
        NSString* category = item[@"category"];
        double price = [item[@"price"] doubleValue];
        NSString* reffer = item[@"reffer"];
        
        YMMECommerceAmount *actualFiat =
                [[YMMECommerceAmount alloc] initWithUnit:@"RUB" value:[[NSDecimalNumber alloc] initWithDouble:price]];
        
        YMMECommercePrice *originalPrice = [[YMMECommercePrice alloc] initWithFiat:actualFiat];
        
        YMMECommerceProduct *product = [[YMMECommerceProduct alloc] initWithSKU:SKU
                                                                           name:name
                                                             categoryComponents:@[category]
                                                                        payload:@{}
                                                                    actualPrice:originalPrice
                                                                  originalPrice:originalPrice
                                                                     promoCodes:@[]];
        
        YMMECommerceReferrer *referrer = [[YMMECommerceReferrer alloc] initWithType:reffer
                                                                         identifier:@""
                                                                             screen:screen];
        
        YMMECommerceCartItem *addedItem = [[YMMECommerceCartItem alloc] initWithProduct:product
                                                                            quantity:quantity
                                                                            revenue:originalPrice
                                                                            referrer:referrer];
        
        [products addObject:addedItem];
    }
    
    
    YMMECommerceOrder *order = [[YMMECommerceOrder alloc] initWithIdentifier:orderId
                                                                   cartItems:products
                                                                     payload:@{}];
    
    [YMMYandexMetrica reportECommerce:[YMMECommerce purchaseEventWithOrder:order] onFailure:nil];
    
    result(nil);
}

@end
